src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,681 @@
+			break;
+		default:
+			ret = -EIO;
+			break;
+		}
+
+		index += param->size;
+	}
+
+	if (ret)
+		comp_err(dev, "waves_codec_configure() error %d", ret);
+	else
+		comp_dbg(dev, "waves_codec_configure() done");
+	return ret;
+}
+
+static int waves_effect_setup_config(struct comp_dev *dev)
+{
+	// apply setup config if present
+	struct codec_data *codec = comp_get_codec(dev);
+	int ret = 0;
+
+	comp_dbg(dev, "waves_effect_setup_config() start");
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
+		comp_err(dev, "waves_effect_startup_config() no setup config");
+		return 0;
+	}
+
+	if (!codec->s_cfg.avail) {
+		comp_warn(dev, "waves_effect_startup_config() no new setup config, using old");
+		codec->s_cfg.avail = true;
+	}
+
+	ret = waves_effect_config(dev, CODEC_CFG_SETUP);
+	codec->s_cfg.avail = false;
+
+	comp_dbg(dev, "waves_effect_setup_config() done");
+	return ret;
+}
+
+int waves_codec_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = NULL;
+	MaxxStatus_t status = 0;
+	int ret = 0;
+
+	comp_dbg(dev, "waves_codec_init() start");
+
+	waves_codec = codec_allocate_memory(dev, sizeof(struct waves_codec_data), 16);
+	if (!waves_codec) {
+		comp_err(dev, "waves_codec_init() allocate memory for waves_codec_data");
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	waves_codec_data_clear(waves_codec);
+	codec->private = waves_codec;
+
+	status = MaxxEffect_GetEffectSize(&waves_codec->effect_size);
+	if (status) {
+		comp_err(dev, "waves_codec_init() MaxxEffect_GetEffectSize() error %d",
+			status);
+		ret = status;
+		codec_free_memory(dev, waves_codec);
+		goto err;
+	}
+
+	waves_codec->effect = (MaxxEffect_t *)codec_allocate_memory(dev,
+		waves_codec->effect_size, 16);
+
+	if (!waves_codec->effect) {
+		comp_err(dev, "waves_codec_init() allocate %d bytes for effect",
+			waves_codec->effect_size);
+		codec_free_memory(dev, waves_codec);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	comp_info(dev, "waves_codec_init() allocated %d bytes for effect",
+		waves_codec->effect_size);
+
+	comp_dbg(dev, "waves_codec_init() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_init() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_prepare(struct comp_dev *dev)
+{
+	int ret = 0;
+
+	comp_dbg(dev, "waves_codec_prepare() start");
+
+	ret = waves_effect_check(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_init(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_allocate(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_setup_config(dev);
+	if (ret)
+		goto err;
+
+	comp_dbg(dev, "waves_codec_prepare() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_prepare() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+
+	comp_dbg(dev, "waves_codec_process() start");
+
+	MaxxStream_t *i_streams[1] = { &waves_codec->i_stream };
```

Comment:
where does 1 come from?

Reply:
this is a number of IO "streams" that a component accepts. 
Added macro for it